projects/RabbitMQ.Client/client/impl/Frame.cs

```diff
@@ -164,96 +152,88 @@ private InboundFrame(FrameType type, int channel, ReadOnlyMemory<byte> payload,
+        private static void ProcessProtocolHeader(ReadOnlySpan<byte> protocolError)
         {
-            try
-            {
-                byte b1 = protocolError.Span[0];
-                byte b2 = protocolError.Span[1];
-                byte b3 = protocolError.Span[2];
-                if (b1 != 'M' || b2 != 'Q' || b3 != 'P')
-                {
-                    throw new MalformedFrameException("Invalid AMQP protocol header from server");
-                }
-
-                int transportHigh = protocolError.Span[3];
-                int transportLow = protocolError.Span[4];
-                int serverMajor = protocolError.Span[5];
-                int serverMinor = await reader.ReadByteAsync().ConfigureAwait(false); ;
-                throw new PacketNotRecognizedException(transportHigh, transportLow, serverMajor, serverMinor);
-            }
-            catch (EndOfStreamException)
+            byte b1 = protocolError[0];
+            byte b2 = protocolError[1];
+            byte b3 = protocolError[2];
+            if (b1 != 'M' || b2 != 'Q' || b3 != 'P')
             {
-                // Ideally we'd wrap the EndOfStreamException in the
-                // MalformedFrameException, but unfortunately the
-                // design of MalformedFrameException's superclass,
-                // ProtocolViolationException, doesn't permit
-                // this. Fortunately, the call stack in the
-                // EndOfStreamException is largely irrelevant at this
-                // point, so can safely be ignored.
                 throw new MalformedFrameException("Invalid AMQP protocol header from server");
             }
+
+            int transportHigh = protocolError[3];
+            int transportLow = protocolError[4];
+            int serverMajor = protocolError[5];
+            int serverMinor = protocolError[6];
+            throw new PacketNotRecognizedException(transportHigh, transportLow, serverMajor, serverMinor);
         }
 
-        internal static async ValueTask<InboundFrame> ReadFrom(PipeReader reader, Memory<byte> frameHeaderBuffer)
+        internal static bool TryReadFrame(ref ReadOnlySequence<byte> buffer, out InboundFrame frame)
         {
-            // We'll always need to read at least 7 bytes (type + channel + payloadSize) or (type + first 6 bytes of protocol error, see ProcessProtocolHeader).
-            int type;
-            try
+            // We'll always need to read at least 8 bytes (type (1) + channel (2) + payloadSize (4) + end marker (1)) or (8 bytes of protocol error, see ProcessProtocolHeader).
+            if (buffer.Length < 8)
             {
-                await reader.ReadAsync(frameHeaderBuffer.Slice(0, 7)).ConfigureAwait(false);
-
-                type = frameHeaderBuffer.Span[0];
+                frame = default;
+                return false;
             }
-            catch (IOException ioe)
+
+            if (buffer.First.Span[0] == 'A')
```

Comment:
accessing chunks this way is  wrong.  just because the ReadOnlySequence conatins 8 bytes the must not be all in the first chunk. the first chunk may onls contain just one byte.

have a look at SequenceReader

Reply:
https://docs.microsoft.com/en-us/dotnet/api/system.buffers.sequencereader-1?view=netcore-3.1

UnreadSpan is the right thing if you want to get the available data that should also eliminate the copies you do below

Reply:
https://github.com/dotnet/runtime/blob/b6d428c3208cf3cc849a5f1f8c77d8663c9de4ea/src/libraries/System.Memory/src/System/Buffers/SequenceReader.cs